### PR TITLE
[MRG] Fix submodule check out

### PR DESCRIPTION
**Pull request by davidbrochart**
This should fix https://github.com/jupyterhub/binderhub/issues/997.
Should I add a test?

### CODE REVIEW

**Comment by betatim** [comment]
We have some tests for the git content provider [here](https://github.com/jupyter/repo2docker/blob/master/tests/unit/contentproviders/test_git.py). It would be good to add a test that makes sure we don't checkout submodules from `master` when using a branch.

**Comment by davidbrochart** [comment]
We can actually clone and check out a specific branch at the same time:
```bash
git clone -b <branch> <remote_repo>
```

**Comment by manics** [comment]
@davidanthoff You can clone branches but not commits https://github.com/jupyterhub/binderhub/issues/997#issuecomment-549453363

**Comment by betatim** [comment]
I think it was for the other David :)

**Comment by manics** [comment]
Sorry, wrong David! No idea why that autocomplete came up, GitHub is normally pretty good about suggesting the most relevant user.

**Comment by davidbrochart** [comment]
The errors in the CI don't seem to be related to my changes.

**Comment by davidbrochart** [comment]
Do you think it can be merged?

**Comment by betatim** [comment]
I'll take a look and the CI jobs have been restarted. (I was waiting for the title to change from WIP to MRG and hence had it on my low priority list.)

File: repo2docker/contentproviders/git.py [hunk 1]
```diff
@@ -21,15 +21,23 @@ def fetch(self, spec, output_dir, yield_output=False):
...
+            else:
+                # don't check out HEAD, the given ref will be checked out later
+                # this prevents HEAD's submodules to be cloned if ref doesn't have them
+                cmd.extend(["--no-checkout"])
             cmd.extend([repo, output_dir])
             for line in execute_cmd(cmd, capture=yield_output):
                 yield line
 
         except subprocess.CalledProcessError as e:
             msg = "Failed to clone repository from {repo}.".format(repo=repo)
```
Comment (betatim) [inline repo2docker/contentproviders/git.py]:
```suggestion
            msg = "Failed to clone repository from {repo}".format(repo=repo)
```

**Comment by betatim** [comment]
The test that fails is passing `origin/b8259dac9eb` as the ref to the checkout command which doesn't work any more so we need to improve on that.

**Comment by betatim** [comment]
This is the first test in https://github.com/jupyter/repo2docker/blob/master/tests/external/reproductions.repos.yaml

You should be able to run it locally with `pytest -svx tests/external/reproductions.repos.yaml::"LIGO Gravitational Waves"` (though I think because the two tests have the same name it might select the "other one" so we need to fix that up as well)

**Comment by betatim** [comment]
Merged! Should get deployed in the next few hours

**Comment by davidbrochart** [comment]
Thanks!

**Pull request merged**